src/meta-srv/src/event_recorder.rs

```diff
@@ -0,0 +1,394 @@
+    /// We use `&[Box<dyn Event>]` to avoid consuming the events, so the caller can buffer the events and retry if the handler fails.
+    async fn handle(&self, events: &[Box<dyn Event>]) -> Result<()>;
+}
+
+/// Implementation of [EventRecorder] that records the events and forwards them to frontend instances for persistence as system tables.
+#[derive(Clone)]
+pub struct EventRecorderImpl {
+    // The channel to send the events to the background processor.
+    tx: Sender<Box<dyn Event>>,
+    // The peer lookup service to fetch the available frontend addresses.
+    peer_lookup_service: PeerLookupServiceRef,
+    // The background processor to process the events.
+    _handle: Option<Arc<JoinHandle<()>>>,
+}
+
+impl EventRecorder for EventRecorderImpl {
+    // Accepts an event and send it to the background handler.
+    fn record(&self, event: Box<dyn Event>) -> Result<()> {
+        self.tx
+            .try_send(event)
+            .map_err(|e| BoxedError::new(PlainError::new(e.to_string(), StatusCode::Internal)))
+            .context(SendEventSnafu)
+    }
+
+    // The default event handler implementation sends the received events to the frontend instances.
+    fn build_event_handler(&self) -> Box<dyn EventHandler> {
+        Box::new(DefaultEventHandlerImpl {
+            peer_lookup_service: self.peer_lookup_service.clone(),
+            channel_manager: ChannelManager::new(),
+        })
+    }
+}
+
+impl EventRecorderImpl {
+    /// Creates a new event recorder to record important events and persist them to the database.
+    pub fn new(meta_peer_client: MetaPeerClientRef) -> Self {
+        let (tx, rx) = channel(DEFAULT_EVENTS_CHANNEL_SIZE);
+        let peer_lookup_service = Arc::new(MetaPeerLookupService::new(meta_peer_client));
+
+        let mut recorder = Self {
+            tx,
+            peer_lookup_service,
+            _handle: None,
+        };
+
+        let processor = EventProcessor::new(rx, recorder.build_event_handler());
+
+        // Spawn a background task to process the events.
+        let handle = tokio::spawn(async move {
+            processor.process().await;
+        });
+
+        recorder._handle = Some(Arc::new(handle));
+
+        recorder
+    }
+}
+
+// DefaultEventHandlerImpl is the default event handler implementation. It sends the received events to the frontend instances.
+struct DefaultEventHandlerImpl {
+    peer_lookup_service: PeerLookupServiceRef,
+    channel_manager: ChannelManager,
+}
+
+#[async_trait]
+impl EventHandler for DefaultEventHandlerImpl {
+    async fn handle(&self, events: &[Box<dyn Event>]) -> Result<()> {
+        debug!("Received {} events: {:?}", events.len(), events);
+        let database_client = self.build_database_client().await?;
+        let row_inserts = RowInsertRequests {
+            inserts: events.iter().map(|e| e.to_row_insert()).collect(),
+        };
```

Comment:
nit: If the events have the same schema, it's more efficient to send them in one insert request.